src/System.Private.CoreLib/shared/System/Number.IntegerFormatting.cs

```diff
@@ -0,0 +1,1542 @@
+                UInt32ToIntegerBuffer(value, ref integer);
+                ValueStringBuilder sb;
+                unsafe
+                {
+                    char* stackPtr = stackalloc char[CharStackBufferSize];
+                    sb = new ValueStringBuilder(new Span<char>(stackPtr, CharStackBufferSize));
+                }
+                if (fmt != 0)
+                {
+                    IntegerBufferToString(ref sb, ref integer, fmt, digits, info);
+                }
+                else
+                {
+                    IntegerBufferToStringFormat(ref sb, ref integer, format, info);
+                }
+                return sb.TryCopyTo(destination, out charsWritten);
+            }
+        }
+
+        public static bool TryFormatUInt64(ulong value, ReadOnlySpan<char> format, IFormatProvider provider, Span<char> destination, out int charsWritten)
+        {
+            // Fast path for default format
+            if (format.Length == 0)
+            {
+                return TryUInt64ToDecStr(value, digits: -1, destination, out charsWritten);
+            }
+
+            char fmt = ParseFormatSpecifier(format, out int digits);
+            char fmtUpper = (char)(fmt & 0xFFDF); // ensure fmt is upper-cased for purposes of comparison
+            if ((fmtUpper == 'G' && digits < 1) || fmtUpper == 'D')
+            {
+                return TryUInt64ToDecStr(value, digits, destination, out charsWritten);
+            }
+            else if (fmtUpper == 'X')
+            {
+                // The fmt-(X-A+10) hack has the effect of dictating whether we produce uppercase or lowercase
+                // hex numbers for a-f. 'X' as the fmt code produces uppercase. 'x' as the format code
+                // produces lowercase.
+                return TryInt64ToHexStr((long)value, (char)(fmt - ('X' - 'A' + 10)), digits, destination, out charsWritten);
+            }
+            else
+            {
+                NumberFormatInfo info = NumberFormatInfo.GetInstance(provider);
+                IntegerBuffer integer = default;
+                UInt64ToIntegerBuffer(value, ref integer);
+                ValueStringBuilder sb;
+                unsafe
+                {
+                    char* stackPtr = stackalloc char[CharStackBufferSize];
+                    sb = new ValueStringBuilder(new Span<char>(stackPtr, CharStackBufferSize));
+                }
+                if (fmt != 0)
+                {
+                    IntegerBufferToString(ref sb, ref integer, fmt, digits, info);
+                }
+                else
+                {
+                    IntegerBufferToStringFormat(ref sb, ref integer, format, info);
+                }
+                return sb.TryCopyTo(destination, out charsWritten);
+            }
+        }
+
+        internal static unsafe void IntegerBufferToString(ref ValueStringBuilder sb, ref IntegerBuffer integer, char format, int nMaxDigits, NumberFormatInfo info)
```

Comment:
As commented [here](https://github.com/dotnet/coreclr/pull/20619#discussion_r228378609), the majority of the code duplication comes from `NumberBufferToString`, `NumberBufferToStringFormat`, and `ParseNumber`.

After looking more closely at the methods, I don't believe there is much that can be done with code sharing. These methods depend on `RoundIntegerBuffer` and `RoundFloatingPointBuffer` on each code path which will end up dealing with the buffers in very different ways. The former thinks in terms of `value`, `scale` and uses "normal" rounding logic; while the latter needs to think in terms of `mantissa`, `exponent`, and uses IEEE-compliant rounding logic (which will ultimately make the end formatting/parsing logic differ between these as well).

However, I think there is a bit of duplication we can remove in other parts of these code files. For example, we have quite a bit of duplication between several of the `X` and `TryX` code paths, namely due to the `X` code path needing to decide if it wants to throw an `OverflowException` or `FormatException` (which could be handled by returning an enum or a secondary bool).

Reply:
Hmmm, actually, we might be able to do it by keeping a single buffer:

```
ref struct NumberBuffer
{
    public int precision;
    public int scale;
    public int exponent; // New field, only used by floating-point buffers
    public bool sign;
    public NumberBufferKind kind;
    public Span<char> digits; // Allocated at creation so it is sized as needed
}
```

The rounding only happens in a few places, at the end of parsing and isn't in the hot-path, so an additional check on `kind` shouldn't hurt. We only need `scale` for integer types, but we can make use of both `scale` and `exponent` for `floating-point` (the former can still be used to help with the fast path check).